### ISSUE TITLE
After upgrade, compactionstats displays odd pending tasks

### ISSUE THREAD

**Issue by powellchristoph** [scylladb/scylla#4228]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 3.0.2-0.20190131.5174b1cd1
Cluster size: 6
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu 1604

I have recently successfully upgrade from 2.0.2-0.20171130.07b039f -> 3.0.2-0.20190131.5174b1cd1. Somewhere along the upgrade process, `nodetool compactionstats` started to display some odd `pending tasks` values. I was previously monitoring and alerting on this value.

A sanitized output displays:
```
pending tasks: 154961134
- example_ks.example_cf: 154961134

id     compaction type keyspace  table      completed total     unit progress
<none> COMPACTION      example_ks example_cf 183842990 338803328 keys 54.26%
Active compaction remaining time :   0h00m00s
```
It will then suddenly drop to `0` with no compactions active. There are no errors in the logs nor any visible impact on the cluster. The column_family schema is very simple and uses the default `'SizeTieredCompactionStrategy`.

I monitored the output for a while and created this gist.
https://gist.github.com/powellchristoph/5f2473bc8865f31fa721f6805171fd8b

**Comment by slivne** [comment]
@amnonh lets start by checking where we pull this info off and if there are no incorrect conversions

**Comment by amnonh** [comment]
The API for pending tasks is:
/compaction_manager/metrics/pending_tasks

I don't think it's an issue, because Cassandra uses int for that as well, is it possible that we'll have wrap-around with the number of pending tasks?

**Comment by slivne** [comment]
we did a small expriment on docker

- ran ```cassandra-stress write n=1000000 -node 172.17.0.2``` in window 1
- killed it after ~1M rows
- did nodetool flush
- ran ```nodetool compactionstats`` ` in window 2

```
[root@e484ce6da61b /]# nodetool compactionstats
pending tasks: 242
- system.compaction_history: 242

id     compaction type keyspace table              completed total unit progress
<none> COMPACTION      system   compaction_history 16        256   keys 6.25%   
<none> COMPACTION      system   compaction_history 14        256   keys 5.47%   
<none> COMPACTION      system   compaction_history 17        256   keys 6.64%   
Active compaction remaining time :   0h00m00s
[root@e484ce6da61b /]# nodetool compactionstats
pending tasks: 239
- system.compaction_history: 239

id     compaction type keyspace table              completed total unit progress
<none> COMPACTION      system   compaction_history 17        256   keys 6.64%   
```

it seems strange ....

**Comment by amnonh** [comment]
The relevant code is in size_tiered_compaction_strategy::estimated_pending_compactions (size_tiered_compaction_strategy.hh)

From looking at the code, is it possible that:
int min_threshold = cf.schema()->min_compaction_threshold();
    int max_threshold = cf.schema()->max_compaction_threshold();

Is not set correctly at an early stage of the node boot?

**Comment by slivne** [comment]
@bhalevy - this is in your turf - @amnonh provided the info from this is retrieved - can you please handle / reassign.

**Comment by bhalevy** [comment]
@raphaelsc can you please look into this?

**Comment by slivne** [comment]
from @glommer 

> I ingested data with cassandra stress, stopped ingesting, and called nodetool compactionstats in a loop. Nothing special

**Comment by raphaelsc** [comment]
Compactionstats was changed to get pending tasks per table using the metric PendingTasksByTableName, but scylla-jmx incorrectly gets an approximation on number of keys left to be compacted instead. scylla-jmx needs to be patched to use "/column_family/metrics/pending_compactions/{name}" for metric PendingTasksByTableName instead.

**Comment by bhalevy** [comment]
> Compactionstats was changed to get pending tasks per table using the metric PendingTasksByTableName, but scylla-jmx incorrectly gets an approximation on number of keys left to be compacted instead. scylla-jmx needs to be patched to use "/column_family/metrics/pending_compactions/{name}" for metric PendingTasksByTableName instead.

@raphaelsc is this something you can do, or do we need to assign someone else?

**Comment by slivne** [comment]
@amnonh back in your court - scylla-jmx change ?

**Comment by raphaelsc** [comment]
@amnonh if you go to scylla-jmx's src/main/java/org/apache/cassandra/metrics/CompactionMetrics.java, you'll see the following:                 map.put(cf, (int)(c.getJsonNumber("total").longValue() - c.getJsonNumber("completed").longValue()));

that's not getting the pending tasks for the column family, that's getting the number of keys left to be compacted.

**Comment by amnonh** [comment]
ok, now I got what you mean, it used to work with: PendingTasks which is correct, but it's no longer the case

**Comment by amnonh** [comment]
BTW, I'm not sure why the compaction struct does not return the real total and completed values

**Comment by amnonh** [comment]
We need Scylla support for this. It's not practical to use the API to query all keyspace/cf sequentially.

**Comment by amnonh** [comment]
With local testing, it works

**Comment by amnonh** [comment]
Should be closed by https://github.com/scylladb/scylla-jmx/commit/71170f5713cdd9436936f539234c27c4511de3ce

**Comment by slivne** [comment]
fixed part of 3.2
